go/logic/migrator.go

```diff
@@ -388,6 +388,12 @@ func (this *Migrator) Migrate() (err error) {
 	if err := this.addDMLEventsListener(); err != nil {
 		return err
 	}
+
+	// write a heartbeat data into the ghc table, and avoid lost data due to MySQL two-phase-commit.
```

Comment:
@shaohk could you add a bit more background in this comment on "why" writing the heartbeat causes us to `avoid lost data due to MySQL two-phase-commit`? That will be useful to those without context

Reply:
I'll try to describe the scene:

When using semi-sync and setting `rpl_semi_sync_master_wait_point=AFTER_SYNC`, if an INSERT statement is being committed but blocks due to an unmet ack count, the data inserted by the transaction is not visible to `ReadMigrationRangeValues`, so the copy of the existing data in the table does not include the new row inserted by the transaction. However, the binlog event for the transaction is already written to the binlog, so the `addDMLEventsListener` only captures the binlog event after the transaction, and thus the transaction's binlog event is not captured, resulting in data loss. 

If you add a heartbeat, and the transaction commit blocks because the ack is not met, then the heartbeat will not be able to write, so the ReadMigrationRangeValues will not be run. When the heartbeat writes successfully, the `ReadMigrationRangeValues` will read the newly inserted data, thus Avoiding data loss due to the above problem

Reply:
@Fanduzi that's my understanding too 👍

I'm hoping we can add some of this context into the code-comment on line 392, so that this is easier to read for those without context

Reply:
@shaohk two more questions:

1. Would it make sense to move this to be within the [`.ReadMigrationRangeValues()`](https://github.com/github/gh-ost/blob/02a26a584dadbe685837ff7d4620134a44196a13/go/logic/applier.go#L424-L433) func?
2. What do you think about updating a [`ChangeLogState`](https://github.com/github/gh-ost/blob/master/go/logic/migrator.go#L24-L29) instead of the heartbeat? Perhaps a new `ChangeLogState` like `ReadMigrationRangeValues` would work here 🤔

Reply:
> @shaohk two more questions:
> 
> 1. Would it make sense to move this to be within the [`.ReadMigrationRangeValues()`](https://github.com/github/gh-ost/blob/02a26a584dadbe685837ff7d4620134a44196a13/go/logic/applier.go#L424-L433) func? The goal with the `.Migrate()` func is to leave a lot of the details to other functions
> 2. What do you think about updating a [`ChangeLogState`](https://github.com/github/gh-ost/blob/master/go/logic/migrator.go#L24-L29) instead of the heartbeat? [`.WriteChangelogState()` in the applier](https://github.com/github/gh-ost/blob/master/go/logic/applier.go#L323-L325) is used for this. Perhaps a new `ChangeLogState` like `ReadMigrationRangeValues` would work here 🤔. [Example PR](https://github.com/github/gh-ost/pull/1134)

Sorry, I just saw these messages.
1. I complete the annotation information in the code.
2. I think the above two suggestions are very good, I will revise the PR and mention it again.

Reply:
> @shaohk two more questions:
> 
> 1. Would it make sense to move this to be within the [`.ReadMigrationRangeValues()`](https://github.com/github/gh-ost/blob/02a26a584dadbe685837ff7d4620134a44196a13/go/logic/applier.go#L424-L433) func? The goal with the `.Migrate()` func is to leave a lot of the details to other functions
> 2. What do you think about updating a [`ChangeLogState`](https://github.com/github/gh-ost/blob/master/go/logic/migrator.go#L24-L29) instead of the heartbeat? [`.WriteChangelogState()` in the applier](https://github.com/github/gh-ost/blob/master/go/logic/applier.go#L323-L325) is used for this. Perhaps a new `ChangeLogState` like `ReadMigrationRangeValues` would work here 🤔. [Example PR](https://github.com/github/gh-ost/pull/1134)

 [Use changelog state instead of the heartbeat](https://github.com/github/gh-ost/pull/1141)

Reply:
> I'll try to describe the scene:
> 
> When using semi-sync and setting `rpl_semi_sync_master_wait_point=AFTER_SYNC`, if an INSERT statement is being committed but blocks due to an unmet ack count, the data inserted by the transaction is not visible to `ReadMigrationRangeValues`, so the copy of the existing data in the table does not include the new row inserted by the transaction. However, the binlog event for the transaction is already written to the binlog, so the `addDMLEventsListener` only captures the binlog event after the transaction, and thus the transaction's binlog event is not captured, resulting in data loss.
> 
> If you add a heartbeat, and the transaction commit blocks because the ack is not met, then the heartbeat will not be able to write, so the ReadMigrationRangeValues will not be run. When the heartbeat writes successfully, the `ReadMigrationRangeValues` will read the newly inserted data, thus Avoiding data loss due to the above problem

Thanks, I added your description to the code comments.  https://github.com/github/gh-ost/pull/1141